File: app/src/main/java/com/example/myanimedata/adapter/AnimeAdapter.java
package com.example.myanimedata.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myanimedata.R;
import com.example.myanimedata.activity.DetailActivity;
import com.example.myanimedata.api.AnimeResult;
import com.makeramen.roundedimageview.RoundedImageView;
import com.squareup.picasso.Picasso;

import java.util.List;

public class AnimeAdapter extends RecyclerView.Adapter<AnimeAdapter.AnimeViewHolder>{
    private final List<AnimeResult> animeResults;
    private final Context context;

    public AnimeAdapter(List<AnimeResult> animeResults, Context context) {
        this.animeResults = animeResults;
        this.context = context;
    }


    @NonNull
    @Override
    public AnimeAdapter.AnimeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new AnimeViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.container_anime, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull AnimeAdapter.AnimeViewHolder holder, int position) {
        holder.bindItem(animeResults.get(position), context);
    }

    @Override
    public int getItemCount() {
        return animeResults.size();
    }

    public static class AnimeViewHolder extends RecyclerView.ViewHolder{
        private final RoundedImageView imageAnimePoster;
        private final TextView textAnimeName;

        public AnimeViewHolder(@NonNull View itemView) {
            super(itemView);

            imageAnimePoster = itemView.findViewById(R.id.imageAnimePoster);
            textAnimeName = itemView.findViewById(R.id.textAnimeName);
        }

        public void bindItem(AnimeResult animeResult, Context context) {

            if(animeResult.getImageResults() != null){
                if(animeResult.getImageResults().getJpgResults() != null){
                    Uri imgUrl = Uri.parse(animeResult.getImageResults().getJpgResults().getImageUrl());
                    Picasso.get().load(imgUrl).into(imageAnimePoster);
                } else if (animeResult.getImageResults().getWebpResults() != null){
                    Uri imgUrl = Uri.parse(animeResult.getImageResults().getWebpResults().getImageUrl());
                    Picasso.get().load(imgUrl).into(imageAnimePoster);
                }
            } else {
                imageAnimePoster.setImageResource(R.drawable.ic_no_image_sm);
            }

            textAnimeName.setText(animeResult.getTitle());

            itemView.setOnClickListener(v ->{
                Intent i = new Intent(context, DetailActivity.class);
                i.putExtra("type", "anime");
                i.putExtra("id", animeResult.getId());
                context.startActivity(i);
            });
        }
    }
}

File: app/src/main/java/com/example/myanimedata/adapter/AnimeAdapter.kt
package com.example.myanimedata.adapter

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.example.myanimedata.R
import com.example.myanimedata.activity.DetailActivity
import com.example.myanimedata.adapter.AnimeAdapter.AnimeViewHolder
import com.example.myanimedata.api.AnimeResult
import com.makeramen.roundedimageview.RoundedImageView
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso

class AnimeAdapter(private val animeResults: List<AnimeResult>, private val context: Context) :
    RecyclerView.Adapter<AnimeViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AnimeViewHolder {
        return AnimeViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.container_anime, parent, false)
        )
    }

    override fun onBindViewHolder(holder: AnimeViewHolder, position: Int) {
        holder.bindItem(animeResults[position], context)
    }

    override fun getItemCount(): Int {
        return animeResults.size
    }

    class AnimeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val loadingAnimeItem: ProgressBar
        private val imageAnimePoster: RoundedImageView
        private val textAnimeName: TextView

        init {
            loadingAnimeItem = itemView.findViewById(R.id.loadingAnimeItem)
            imageAnimePoster = itemView.findViewById(R.id.imageAnimePoster)
            textAnimeName = itemView.findViewById(R.id.textAnimeName)
        }

        fun bindItem(animeResult: AnimeResult, context: Context) {
            loadingAnimeItem.visibility = View.GONE
            val imgUrl = Uri.parse(animeResult.imageResults.jpgResults.imageUrl)
            Picasso.get().load(imgUrl).noFade().into(imageAnimePoster, object : Callback {
                override fun onSuccess() {
                    imageAnimePoster.animate().setDuration(500).alpha(1f).start()
                }

                override fun onError(e: Exception) {
                    Toast.makeText(
                        itemView.context, e.message + " cause : "
                                + e.cause, Toast.LENGTH_SHORT
                    ).show()
                }
            })
            textAnimeName.text = animeResult.title
            itemView.setOnClickListener {
                val i = Intent(context, DetailActivity::class.java)
                i.putExtra("type", "anime")
                i.putExtra("id", animeResult.id)
                context.startActivity(i)
            }
        }
    }
}